### PR TITLE
Improve response handling - Closes #410

### LINKED ISSUE #410: Improve response handling
### Description

Improve response handling so it reflects the specifications attached below. 

### Motivation

- Proper exception handling combined with correct HTTP error codes is important to inform user about input errors and service outages

### Acceptance Criteria

- Gateway is able to process various error scenarios and report it properly in the response
- Gateway covers nicely the stack trace and internal error messages
- It works well with HTTP and RPC (mind the different error codes in RPC)
- All integration test cases must also be adjusted to be inline with the changes

### Additional Information

Specification: https://github.com/LiskHQ/lisk-service/blob/development/docs/api/version2.md

### PR DESCRIPTION
**Pull request by nagdahimanshu**
### What was the problem?
This PR resolves #410 

### How was it solved?

- [x] Gateway is able to process empty results with 200 OK
- [x] It works well with HTTP 
- [x] All integration test cases must also be adjusted to be inline with the changes

### How was it tested?
Local

### CODE REVIEW

File: services/blockchain-indexer/shared/dataService/business/blocks.js [hunk 1]
```diff
@@ -218,10 +217,7 @@ const getBlocks = async params => {
 	} catch (err) {
 		// Block does not exist
 		if (err.message.includes('does not exist')) {
-			let errMessage;
-			if ('id' in params && err.message.includes(params.id)) errMessage = `Block with ID ${params.id} does not exist`;
-			if ('height' in params) errMessage = `Block at height ${params.height} does not exist (${err.message})`;
-			throw new NotFoundException(errMessage || err.message);
+			return blocks;
```
Comment (priojeetpriyom) [inline services/blockchain-indexer/shared/dataService/business/blocks.js:220]: Will execution come to this condition anymore as we `throw` statements with `return`?

File: services/blockchain-indexer/methods/dataService/controllers/transactions.js [hunk 2]
```diff
@@ -73,7 +67,6 @@ const getTransactions = async (params) => {
 		let status;
 		if (err instanceof InvalidParamsException) status = 'INVALID_PARAMS';
 		if (err instanceof ValidationException) status = BAD_REQUEST;
-		if (err instanceof NotFoundException) status = NOT_FOUND;
 		if (err.message.includes('does not exist')) status = NOT_FOUND;
```
Comment (sameersubudhi) [inline services/blockchain-indexer/methods/dataService/controllers/transactions.js:70]: I think, this line also needs to go. Could you please double check?
```suggestion
```

File: services/blockchain-indexer/shared/dataService/business/blocks.js [hunk 3]
```diff
@@ -193,15 +192,15 @@ const getBlocks = async params => {
 
 	try {
 		if (params.ids) {
-			if (Array.isArray(params.ids) && !params.ids.length) throw new NotFoundException(`Block with IDs: ${params.ids.join(', ')} not found.`);
+			if (Array.isArray(params.ids) && !params.ids.length) return blocks;
```
Comment (sameersubudhi) [inline services/blockchain-indexer/shared/dataService/business/blocks.js:195]: We can probably remove these checks here. Please verify.

File: services/blockchain-indexer/shared/dataService/business/blocks.js [hunk 4]
```diff
@@ -193,15 +192,15 @@ const getBlocks = async params => {
...
 			blocks.data = await getBlocksByIDs(params.ids);
 		} else if (params.id) {
 			blocks.data.push(await getBlockByID(params.id));
-			if (Array.isArray(blocks.data) && !blocks.data.length) throw new NotFoundException(`Block ID ${params.id} not found.`);
+			if (Array.isArray(blocks.data) && !blocks.data.length) return blocks;
```
Comment (sameersubudhi) [inline services/blockchain-indexer/shared/dataService/business/blocks.js:199]: Same as above.

File: services/blockchain-indexer/shared/dataService/business/blocks.js [hunk 5]
```diff
@@ -193,15 +192,15 @@ const getBlocks = async params => {
...
 			if ('offset' in params && params.limit) blocks.data = blocks.data.slice(params.offset, params.offset + params.limit);
 		} else if (params.height) {
 			blocks.data.push(await getBlockByHeight(Number(params.height)));
-			if (Array.isArray(blocks.data) && !blocks.data.length) throw new NotFoundException(`Height ${params.height} not found.`);
+			if (Array.isArray(blocks.data) && !blocks.data.length) return blocks;
```
Comment (sameersubudhi) [inline services/blockchain-indexer/shared/dataService/business/blocks.js:203]: Same as above.

File: services/blockchain-indexer/shared/dataService/business/blocks.js (hunk 1 above)
Comment (sameersubudhi) [reply services/blockchain-indexer/shared/dataService/business/blocks.js:220]: @priojeetpriyom Your question is a bit unclear. Could you please clarify?

@nagdahimanshu We can probably replace the entire catch body with the following statement:
```
if (!err.message.includes('does not exist')) throw err;
```

File: tests/schemas/api_v3/block.schema.js [hunk 6]
```diff
@@ -14,17 +14,22 @@
  *
  */
 import Joi from 'joi';
+import regex from './regex';
+
+const generator = {
+	address: Joi.string().pattern(regex.ADDRESS_LISK32).required(),
```
Comment (sameersubudhi) [inline tests/schemas/api_v3/block.schema.js:20]: Is it possible to resolve the generator `publicKey` and `name` to extend this response?

File: tests/schemas/api_v3/block.schema.js [hunk 7]
```diff
@@ -14,17 +14,22 @@
...
-	assetsRoot: Joi.string().required(),
-	stateRoot: Joi.string().required(),
-	transactionRoot: Joi.string().required(),
-	previousBlockID: Joi.string().required(),
+	generator: Joi.object(generator).required(),
+	assetRoot: Joi.string().pattern(regex.HASH_SHA256).required(),
+	stateRoot: Joi.string().pattern(regex.HASH_SHA256).required(),
+	transactionRoot: Joi.string().pattern(regex.HASH_SHA256).required(),
+	previousBlockID: Joi.string().pattern(regex.HASH_SHA256).required(),
 	signature: Joi.string().allow('').required(),
```
Comment (sameersubudhi) [inline tests/schemas/api_v3/block.schema.js:33]: Could we please include the pattern check for `signature` too?

File: tests/schemas/api_v3/block.schema.js [hunk 8]
```diff
@@ -14,17 +14,22 @@
...
 	aggregateCommit: Joi.object().required(),
```
Comment (sameersubudhi) [inline tests/schemas/api_v3/block.schema.js:34]: Is it possible to specify the object structure for `aggregateCommit` here? Kindly ensure the same is reflected in the gateway source mappings.

File: tests/schemas/api_v3/block.schema.js [hunk 9]
```diff
@@ -34,7 +39,10 @@ const blockSchema = {
 	totalBurnt: Joi.string().required(),
 	maxHeightGenerated: Joi.number().required(),
 	maxHeightPrevoted: Joi.number().required(),
-	validatorsHash: Joi.string().required(),
+	validatorsHash: Joi.string().pattern(regex.HASH_SHA256).required(),
+	numberOfTransactions: Joi.number().integer().min(0).required(),
+	numberOfAssets: Joi.number().integer().min(1).required(),
+	numberOfEvents: Joi.number().integer().min(1).required(),
```
Comment (sameersubudhi) [inline tests/schemas/api_v3/block.schema.js:45]: Please confirm with the SDK team if there are scenarios where `numberOfAssets` and `numberOfEvents` could be `0`? If so, please adjust accordingly.

File: tests/schemas/api_v3/delegate.schema.js [hunk 10]
```diff
@@ -34,15 +34,15 @@ const pomHeightsSchema = {
...
 	voteWeight: Joi.string().pattern(regex.VOTE_WEIGHT).required(),
 	address: Joi.string().pattern(regex.ADDRESS_LISK32).required(),
-	lastGeneratedHeight: Joi.number().integer().positive().required(),
+	lastGeneratedHeight: Joi.number().integer().min(0).required(),
 	status: Joi.string().valid(...DELEGATE_STATUSES).required(),
 	isBanned: Joi.boolean().required(),
 	pomHeights: Joi.array().items(pomHeightsSchema).required(),
-	consecutiveMissedBlocks: Joi.number().integer().positive().min(0)
+	consecutiveMissedBlocks: Joi.number().integer().min(0)
 		.required(),
```
Comment (sameersubudhi) [inline tests/schemas/api_v3/delegate.schema.js:46]:
```suggestion
	consecutiveMissedBlocks: Joi.number().integer().min(0).required(),
```

**Review by sameersubudhi — CHANGES_REQUESTED** [review]

File: tests/schemas/api_v3/block.schema.js (hunk 6 above)
Comment (nagdahimanshu) [reply tests/schemas/api_v3/block.schema.js:20]: Yes, it is already added. Just need to add those in test schemas.

**Review by priojeetpriyom — APPROVED** [review]

**Review by sameersubudhi — APPROVED** [review]

**Pull request merged**